ci: add android and compose lint [WPB-9287]

vitorhugods opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9287" title="WPB-9287" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-9287</a>  Compose Detekt Rules
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [X] contains a reference JIRA issue number like `SQPIT-764`
    - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

There are many localisation, performance and other hidden bugs in our app that we aren't aware of.

### Solutions

We could start by adding Lint checks. Extra: adding [compose lint checks](https://slackhq.github.io/compose-lints/).

Add these check to the codestyle workflow to catch new smelly stuff as PRs are raised.

For now, I've added all the current bugs to the baseline files, generated with run `./gradlew updateLintBaseline`.

The lint configuration is done in `configureKotlinAndroid` function, shared by all of our Android gradle modules.

### Testing

Just run `./gradlew lint` on the root of the project.

### Attachments

When running locally, we can also open the HTML file for easier reading. This is what is currently seen if we take out the baseline file:

![image](https://github.com/wireapp/wire-android/assets/9389043/814a41a3-4139-4cef-918d-651398af0371)

Biggest offender is the `ImpliedQuantity` check for localisation:

![image](https://github.com/wireapp/wire-android/assets/9389043/01b4a0c2-4162-47d5-92d0-c45228a92ff2)

----
#### PR Post Merge Checklist for internal contributors

- [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

vitorhugods committed:
ci: add android and compose lint

vitorhugods committed:
ci: rename detekt to code-style

vitorhugods committed:
chore: update aboutLibraries to 11.1.4

AndroidBob commented:
Build [4860](https://10.10.124.134/job/AR-build-pipeline/4860/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9733527)(9733527)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9733527)(9733527)-staging-compat.apk)

Garzas reviewed (commented):
Code looks good but now build will take longer and I'm thinking how to solve that 🤔 
(Lint file has 30MB of logs)

vitorhugods committed:
ci: run lintDebug instead of just lint

AndroidBob commented:
Build [4864](https://10.10.124.134/job/AR-build-pipeline/4864/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9734090)(9734090)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9734090)(9734090)-staging-compat.apk)

vitorhugods committed:
ci: Increase Metaspace size in GitHub Actions workflow

The Metaspace size parameter has been increased in the codestyle.yml GitHub Actions workflow file. This was just to prevent the linter runs from failing on the GH runner due to limited Metaspace size. The value of the MaxMetaspaceSize option is set to 1g.

vitorhugods commented:
> Code looks good but now build will take longer and I'm thinking how to solve that 🤔 (Lint file has 30MB of logs)

The logs seem triggered by [AboutLibraries](https://github.com/mikepenz/AboutLibraries). Maybe we can go around it someway.

The main issue seems unrelated to that, as more people seem to face issues with GH Actions + Lint: https://stackoverflow.com/a/70010526

AndroidBob commented:
Build [4868](https://10.10.124.134/job/AR-build-pipeline/4868/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9735613)(9735613)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9735613)(9735613)-staging-compat.apk)

Garzas commented on app/lint-baseline.xml:
As I understand those baseline's were added to in future remove those ignores and for future releases already check new places where we may have those issues?

vitorhugods replied:
Yes. Using `lintDebug` was skipping `:app`'s lint. Now it's fixed.

vitorhugods committed:
ci: update linter command in GitHub actions workflow

vitorhugods committed:
ci: fix comment breaking the lint commands

vitorhugods committed:
ci: silence dependencies collection in GitHub Workflow

AndroidBob commented:
Build [4879](https://10.10.124.134/job/AR-build-pipeline/4879/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9740365)(9740365)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9740365)(9740365)-staging-compat.apk)

vitorhugods committed:
ci: disable daemon for debugging

vitorhugods committed:
ci: disable AboutLibraries when running lint checks

vitorhugods committed:
ci: run style and tests in parallel

vitorhugods committed:
fix: wrong string argument type

AndroidBob commented:
Build [4893](https://10.10.124.134/job/AR-build-pipeline/4893/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9741297)(9741297)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9741297)(9741297)-staging-compat.apk)

yamilmedina reviewed (commented):
Looks fantastic, thanks for picking up an old collective topic 🥇 

One question, how this lint vs. a [detekt plugin for compose](https://detekt.dev/marketplace) for [example](https://github.com/appKODE/detekt-rules-compose) differs performance wise or in terms of maintainability ? 

Maybe we could be better off reusing the same tool, in this case `detekt`, than introducing a new one 🤔

vitorhugods commented:
> Looks fantastic, thanks for picking up an old collective topic 🥇
> 
> One question, how this lint vs. a [detekt plugin for compose](https://detekt.dev/marketplace) for [example](https://github.com/appKODE/detekt-rules-compose) differs performance wise or in terms of maintainability ?
> 
> Maybe we could be better off reusing the same tool, in this case `detekt`, than introducing a new one 🤔

Maybe we can split by parts.

This PR adds two things: **Android Lint**, and **Compose checks for Android Lint**.

### Android Lint

Is the heavier guy here in terms of performance. Unlike Detekt, it will compile the code and check the compiler output for weird things. Because of that, it can analyse more stuff (like class structures) and even analyse resource files and catch localisation issues.

I believe adding Android Lint in general is beneficial. It just noticed we have 200+ localisation issues, for example.

### Compose Checks for Android Lint

It's an extra check. By itself doesn't add much to verification time.

But continuing on your question: should it be on top of Detekt, or on top of Android Lint?

It seems some of these Detekt plugins rely on Detekt's [Type Resolution](https://detekt.dev/docs/gettingstarted/type-resolution/), which we don't have set-up at the moment. And this would mean spinning up the Kotlin Compiler as well for Detekt, which we don't at the moment.
Some of them don't rely on that, so I'm not 100% sure they would be able to cover as much as the others.

The Detekt rulesets for Compose I found were not as popular and actively maintained as this Android Lint alternative by Slack. These aren't the only things to consider, but when selecting such a delicate topic, I'd go with the more commonly used tool.

yamilmedina commented on .github/workflows/codestyle.yml:
```diff
@@ -25,6 +33,9 @@ jobs:
             ./gradlew detektAll
+      - name: Run Linter
+        run: |
+            ./gradlew lint
```
**Suggestion**: I think, it makes sense to create a job out of this, instead of sequential steps, so might be able to run these 2 in parallel.

vitorhugods replied:
It does have the drawback that it will setup java and download dependencies twice in parallel, so the benefit is smaller.

But we do get lint / codestyle results faster.

yamilmedina reviewed (commented):
Thanks for the replies, it makes sense, just added another suggestion we might explore before merging, to see how much (if any) time could gain of execution

vitorhugods committed:
ci: run code style and lint in parallel

vitorhugods committed:
ci: only build of code-analysis succeeds

yamilmedina reviewed (approved):
Coool 🚀

AndroidBob commented:
Build [4911](https://10.10.124.134/job/AR-build-pipeline/4911/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9742291)(9742291)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9742291)(9742291)-staging-compat.apk)
